ostia-operator/pkg/apicast/apicast.go

```diff
@@ -238,39 +226,145 @@ func newProbe(path string, port int32, initDelay int32, timeout int32, period in
 	}
 }
 
-//TODO - This should returna slice of generic "RateLimit" objects of multiple, mixed types
-func buildRateLimitPolicies(validSvcs []string, limits []v1alpha1.RateLimit) []FixedWindowRateLimiter {
-	var fixedLimitRules []FixedWindowRateLimiter
-	for _, v := range limits {
-		if contains(validSvcs, v.Service) {
-			switch v.Type {
-			case "fixed":
-				if v.TimeUnit == "minutes" {
-					v.TimeInterval = v.TimeInterval * 60
-				}
-				rule := FixedWindowRateLimiter{
-					Count:  v.MaxRequest,
-					Window: v.TimeInterval,
-					Key: LimiterKey{
-						Name:  v.Service,
-						Scope: v.Service,
-					},
-				}
-				fixedLimitRules = append(fixedLimitRules, rule)
-			default:
-				log.Errorf("unknown rate limit type %s for service. ignoring", v.Type, v.Service)
-			}
+func processRateLimitPolicies(limits []v1alpha1.RateLimit) PolicyChain {
+	var fixedLimiters []FixedWindowRateLimiter
+	var leakyLimiters []LeakyBucketRateLimiter
+	var connLimiters []ConnectionRateLimiter
+
+	for _, limit := range limits {
+		rl, err := parseRateLimit(limit)
+		if err != nil {
+			log.Error(fmt.Sprintf("error processing rate limit rule for %s - %s", limit.Name, err.Error()))
+			continue
 		}
 
+		switch rl.(type) {
+		case FixedWindowRateLimiter:
+			fwRl := rl.(FixedWindowRateLimiter)
+			fixedLimiters = append(fixedLimiters, fwRl)
+		case LeakyBucketRateLimiter:
+			lbRl := rl.(LeakyBucketRateLimiter)
+			leakyLimiters = append(leakyLimiters, lbRl)
+		case ConnectionRateLimiter:
+			crRl := rl.(ConnectionRateLimiter)
+			connLimiters = append(connLimiters, crRl)
+		default:
+			log.Errorf("unknown rate limit type %T. ignoring", rl)
+			continue
+		}
 	}
-	return fixedLimitRules
+
+	var config PolicyChainConfiguration
+	if len(fixedLimiters) > 0 {
+		config.FixedWindowLimiters = &fixedLimiters
+	}
+	if len(leakyLimiters) > 0 {
+		config.LeakyBucketLimiters = &leakyLimiters
+	}
+	if len(connLimiters) > 0 {
+		config.ConnectionLimiters = &connLimiters
+	}
+
+	return PolicyChain{"apicast.policy.rate_limit", config}
 }
 
-func contains(serviceNames []string, svc string) bool {
-	for _, service := range serviceNames {
-		if service == svc {
-			return true
+func parseRateLimit(rl v1alpha1.RateLimit) (interface{}, error) {
+	key := LimiterKey{rl.Name, "plain", "service"}
+	if rl.Source != "" {
+		switch rl.Source {
+		//TODO - Add support for more sources here, jwt etc ..
```

Comment:
I'm not sure this can be abstracted. JWT has many claims and keys can contain combinations of them.

Reply:
Again to explain this, there was a comment about not tying the CRD to apicast config. I wasn't aware there would be so many options just going from what I could find. Whats the best approach here? Should we give the user (via docs or examples) a list of valid inputs and have them pass it in explicitly?

I think that approach is ok personally.

Reply:
There is infinite amount of valid inputs. JWT can contain anything. It is arbitrary hash structure.

IMO we have to figure out how to give people the flexibility of defining keys of dynamic data. For example any http header. And combinations of them like described in 
https://github.com/3scale/ostia/pull/20#discussion_r207470054.